### PR TITLE
more methods in imagegen

### PR DESCRIPTION
**Pull request by yurkor**
I propose to add new methods to ImageDataGenerator.
The problem you can not use same generator for train, test and validation data.

PS This is my first merge request to open source project and I do not code in python usually.

### CODE REVIEW

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -437,6 +437,50 @@ def fit(self, X,
...
+
+        Arguments:
+            x: image(Numpy array) to transform
+            train_mode: True if random augmentation should be applied, False - steps needed to be applied on tst and train data(aka normalisation)
+
+        Returns: transformed image
+        '''
+        if train_mode:
+            x = self.random_transform(x)
+        return self.standardize(x)
```
Comment (fchollet) [inline keras/preprocessing/image.py]: I don't understand why you need a new method for this. Surely calling `random_transform` and/or `standardize` should be enough?

Comment (yurkor) [inline keras/preprocessing/image.py]: I did it to better match common use case.
User only need both(transform+ sstandartize) or only standartize(for testing). And suggest to make other methods private as they can lead to mistake .

Comment (fchollet) [inline keras/preprocessing/image.py]: It doesn't really make sense to me, because random transformations are important to have at test time as well (because they modify the statistics of the data). Best evals are from multiple random transforms, with results merged via power averaging.

Comment (yurkor) [inline keras/preprocessing/image.py]: I am not that experienced in image processing(thanks to this library I understand some cool DL concepts and image ideas). I could miss something but after reading some papers  about building models such as VGG16/18 I got idea that you do random transformation to artificially increase number of training examples and therefore prevent overfitting and have ability to use more complex models.  So they usually do random transformation for training examples and apply normalization/standardization to eval and test data. I guess you right that using random transformation on eval and testing is common, but I also think it is common not to do them on test data. And data preprocessing is always have to be applied to train and test, because you don't want train on mean subtracted data and apply it to test. 

I am thinking about something like transformers in sklearn, but with DL specific you might want to have slightly different transformation on test/train. And with two options train_mode=True/False it cover all use cases.

Comment (yurkor) [inline keras/preprocessing/image.py]: what if I change train_mode to random_transform? Or you think there should not be any case where you need only one of them. I think single mode, batch mode and generator mode still can  be useful with/without train_mode(or other name) param.

**Comment by joelthchao** [comment]
I don't understand your scenario very much. If you want to share same generator with same random transform setting for different set of data, my workaround is create new generator with same setting and fit it to the same data. Of course, you can manually setup `self.mean` / `self.std` / `self.principle_components` and skip the fitting step.

Probably, only thing you need to write is a util method to copy an old generator?

**Comment by yurkor** [comment]
You probably right and util would be better place for these methods. Not sure if it is good idea to run fit multiple times as if it might have some randomized algorithm in future(like randomized PCA etc).

Just giving my example
 `
train_raw, train_Y, valid_raw, valid_Y, test_raw = ...
datagen = ImageDataGenerator(...)
datagen.fit(train_raw)
model.fit_generator(
gen.generator(train_raw,valid_Y),
samples_per_epoch=len(train_raw), 
validation_data=(datagen.batch_transform(valid_raw, False), valid_Y))

model.predict_class(datagen.batch_transform(test_raw))
`

It could be more and more useful as there are more methods in Generator and you have to apply them in right order.

**Comment by joelthchao** [comment]
Another way to achieve this is making `ImageDataGenerator.flow` always produce a new generator when it is called multiple time.
```python
datagen = ImageDataGenerator(...) # normalization, standardization, rescale + random augmentation
datagen.fit(train_X)

model.fit_generator(
    datagen.flow(train_X, train_Y),
    samples_per_epoch=len(train_X), 
    validation_data=(datagen.flow(valid_X, valid_Y)))

model.predict_class(datagen.flow(test_X, test_Y))
```

**Comment by yurkor** [comment]
If my understanding correct, flow does not produce multiple generators it return self(and modifies state) and class itself is an Iterator.

**Comment by joelthchao** [comment]
Of course, I mean to modify it into producing multiple generators.

**Comment by fchollet** [comment]
There is one issue with creating a new generator everytime `flow` is called: keeping the same generator allows us to keep looping over the data in order even if each epoch does not look at all the samples. So if you have 10M samples, and your epochs are 100k samples, by keeping the same generator everytime `flow` is called, `fit_on_generator` will be able to loop over all the data in batches of 100k. If independent generators were created, you would always loop on the first 100k samples.

Why couldn't you instead create a copy of a generator and reinitialize it?

**Comment by yurkor** [comment]
But if you want to keep same generator why not to store it in variable and reuse where it left off?
i.e. instead of `model.fit_generator(datagen.flow(X,y),epoch=1)` do 
`
gen =datagen.flow(X,y);
model.fit_generator(gen,epoch=1);
model.fit_generator(gen,epoch=1);
`
